File: frameworks/java/vertx/UnitTests/src/test/java/unittests/StreamTests.java
package unittests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.net.http.HttpResponse;

import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

class StreamTests extends VertxApp {

    // rb:test stream.ndjson
    @Test
    @Tag("stream.ndjson")
    void eachRowIsALineAndTheLengthIsNeverSent() throws Exception {
        HttpResponse<byte[]> response = get("/stream/items");

        assertTrue(Answer.header(response, "content-type").startsWith("application/x-ndjson"));
        assertNull(Answer.header(response, "content-length"));
        String[] lines = Answer.text(response).stripTrailing().split("\n");
        JsonArray rows = Expected.json("items.medium.json").getJsonArray("items");
        assertEquals(rows.size(), lines.length);
        for (int i = 0; i < lines.length; i++) {
            assertEquals(rows.getJsonObject(i), new JsonObject(lines[i]));
        }
    }
}

File: frameworks/java/ktor/UnitTests/src/test/kotlin/unittests/StreamTests.kt
package unittests

import io.ktor.client.request.get
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpStatusCode
import io.ktor.http.contentType
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.jsonObject
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test

class StreamTests {
    // rb:test stream.ndjson
    @Test
    @Tag("stream.ndjson")
    fun `each row of items medium is a line with no length`() = corpusTest {
        val response = client.get("/stream/items")

        assertEquals(HttpStatusCode.OK, response.status)
        assertEquals(ContentType("application", "x-ndjson"), response.contentType()?.withoutParameters())
        assertNull(response.headers[HttpHeaders.ContentLength])
        val rows = response.bodyAsText().lines().filter { it.isNotEmpty() }.map(Json::parseToJsonElement)
        assertEquals(Expected.json("items.medium.json").jsonObject["items"], JsonArray(rows))
    }
}
